### ISSUE TITLE
Cannot use generic proc or iterator from forward-invoked generic proc or iterator

### ISSUE THREAD

**Issue by endragor** [nim-lang/Nim#4820]
Sample:

``` nim
proc printSeq[T](s: seq[T])
printSeq(@[1])

proc printSeq[T](s: seq[T]) =
  for item in items(s):
    echo item
```

Compile-time error:

```
sample.nim(5, 15) Error: type mismatch: got (seq[T])
but expected one of:
iterator items[IX, T](a: array[IX, T]): T
iterator items[](E: typedesc[enum]): E:type
iterator items[T](s: Slice[T]): T
iterator items(a: string): char
iterator items[T](a: seq[T]): T
iterator items[T](a: openArray[T]): T
iterator items[T](a: set[T]): T
iterator items(a: cstring): char
```

If the `printSeq` proc gets invoked only after the definition, then this compiles fine.
Nim Git hash: 20b844ec53714170957abce87d67720cd1f29f22

**Comment by endragor** [comment]
Hmm, this might be a duplicate of #4104 that I filed several months ago, but the symptom here is slightly different. And it seems #4104 now produces an error similar to this one. Any chance this could be fixed quickly? We use forward declarations extensively in certain places due to lack of package-visibility or circular-import support.

**Comment by Araq** [comment]
Quickly? I think I can fix it this weekend.

**Comment by endragor** [comment]
I meant "easy to fix". Thank you, but this is not critical, actually. I restructured code a bit so we're fine until more generic stuff appears :)

**Comment by Yardanico** [comment]
@Araq fixed in devel

**Comment by Yardanico** [comment]
Fixed in https://github.com/nim-lang/Nim/commit/be174fc3c731f1aecf07d7750c038dbb7a018812, test-case is already there - https://github.com/nim-lang/Nim/blob/be174fc3c731f1aecf07d7750c038dbb7a018812/tests/generics/tforwardgeneric.nim#L20
